### PR TITLE
Version 4.0

### PR DESCRIPTION
**Pull request by willfarrell**
See https://github.com/middyjs/middy/issues/769 for details

- No breaking changes to existing 3rd party middlewares
- Release blocked by: https://github.com/aws/aws-lambda-base-images/issues/47

Closes:
- v4 roadmap #769
- normalize statusCode #874

### CODE REVIEW

File: website/docs/middlewares/ws-response.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ npm install --save @middy/ws-responder
 ```
 
 ## Options
-- `AwsClient` (object) (default `AWS.ApiGatewayManagementApi`): AWS.ApiGatewayManagementApi class constructor (e.g. that has been instrumented with AWS XRay). Must be from `aws-sdk` v2.
+- `AwsClient` (object) (default `ApiGatewayManagementApiClient`): ApiGatewayManagementApi class constructor (e.g. that has been instrumented with AWS XRay). Must be from `@aws-sdk/client-apigatewaymanagementapi`.
```
Comment (naorpeled) [inline website/docs/middlewares/ws-response.md:16]:
```suggestion
- `AwsClient` (object) (default `ApiGatewayManagementApiClient`): ApiGatewayManagementApi class constructor (i.e. that has been instrumented with AWS XRay). Must be from `@aws-sdk/client-apigatewaymanagementapi`.
```

I think this fits more, wdyt?

File: packages/util/__tests__/index.js [hunk 2]
```diff
@@ -485,43 +489,45 @@ test('normalizeHttpResponse should not change response', async (t) => {
...
 })
 
 test('normalizeHttpResponse should update undefined response', async (t) => {
   const request = {}
   normalizeHttpResponse(request)
-  t.deepEqual(request, { response: { headers: {} } })
+  t.deepEqual(request, { response: { statusCode: 500, headers: {} } })
 })
 
 test('normalizeHttpResponse should update nullish response', async (t) => {
```
Comment (naorpeled) [inline packages/util/__tests__/index.js:508]: not sure if I'm missing it, but let's add a test that asserts the response if only part of the response params are presnet (e.g. only body or only status code)

File: packages/sts/index.d.ts [hunk 3]
```diff
@@ -1,11 +1,11 @@
 import middy from '@middy/core'
 import { Options as MiddyOptions } from '@middy/util'
 import { Context as LambdaContext } from 'aws-lambda'
-import STS from 'aws-sdk/clients/sts'
+import { STSClient } from '@aws-sdk/client-sts'
 
-interface Options<S = STS>
+interface Options<S = STSClient>
```
Comment (naorpeled) [inline packages/sts/index.d.ts:6]: let's improve the naming here and rename S to something that resembles the actual meaning of it more.

File: packages/ssm/index.js [hunk 4]
```diff
@@ -9,12 +9,15 @@ import {
...
+import {
+  SSMClient,
+  GetParametersCommand,
+  GetParametersByPathCommand
+} from '@aws-sdk/client-ssm'
 
 const awsRequestLimit = 10
 const defaults = {
-  AwsClient: SSM, // Allow for XRay
+  AwsClient: SSMClient, // Allow for XRay
```
Comment (naorpeled) [inline packages/ssm/index.js:20]: Why do we need this for XRay? (just wondering)

File: packages/ssm/index.d.ts [hunk 5]
```diff
@@ -1,11 +1,11 @@
 import middy from '@middy/core'
 import { Options as MiddyOptions } from '@middy/util'
 import { Context as LambdaContext } from 'aws-lambda'
-import SSM from 'aws-sdk/clients/ssm'
+import { SSMClient } from '@aws-sdk/client-ssm'
 import { JsonValue } from 'type-fest'
 
-interface Options<S = SSM>
-  extends MiddyOptions<S, SSM.Types.ClientConfiguration> {}
+interface Options<S = SSMClient>
```
Comment (naorpeled) [inline packages/ssm/index.d.ts:7]: let's rename `S` to something that resembles its actual meaning more.

File: packages/sqs-partial-batch-failure/index.d.ts [hunk 6]
```diff
@@ -1,8 +1,16 @@
 import middy from '@middy/core'
+import { SQSClient } from '@aws-sdk/client-sqs'
+import { Options as MiddyOptions } from '@middy/util'
 
-interface Options {
-  logger?: (reason: any, record: any) => void
-}
+interface Options<S = SQSClient>
```
Comment (naorpeled) [inline packages/sqs-partial-batch-failure/index.d.ts:5]:
```suggestion
interface Options<AwsSQSClient = SQSClient>
```

File: packages/service-discovery/index.d.ts [hunk 7]
```diff
@@ -1,13 +1,14 @@
...
 import { Context as LambdaContext } from 'aws-lambda'
-import ServiceDiscovery, {
+import {
+  ServiceDiscoveryClient,
   HttpInstanceSummaryList
-} from 'aws-sdk/clients/servicediscovery'
+} from '@aws-sdk/client-servicediscovery'
 
-interface Options<S = ServiceDiscovery>
+interface Options<S = ServiceDiscoveryClient>
```
Comment (naorpeled) [inline packages/service-discovery/index.d.ts:9]:
```suggestion
interface Options<AwsServiceDiscoveryClient = ServiceDiscoveryClient>
```

File: packages/secrets-manager/index.d.ts [hunk 8]
```diff
@@ -1,10 +1,10 @@
 import middy from '@middy/core'
 import { Options as MiddyOptions } from '@middy/util'
 import { Context as LambdaContext } from 'aws-lambda'
-import SecretsManager from 'aws-sdk/clients/secretsmanager'
+import { SecretsManagerClient } from '@aws-sdk/client-secretsmanager'
 
-interface Options<SM = SecretsManager>
-  extends MiddyOptions<SM, SecretsManager.Types.ClientConfiguration> {}
+interface Options<SM = SecretsManagerClient>
+  extends MiddyOptions<SM, SecretsManagerClient.Types.ClientConfiguration> {}
```
Comment (naorpeled) [inline packages/secrets-manager/index.d.ts:7]:
```suggestion
interface Options<AwsSecretsManagerClient= SecretsManagerClient>
  extends MiddyOptions<AwsSecretsManagerClient, SecretsManagerClient.Types.ClientConfiguration> {}
```

File: packages/s3-object-response/index.d.ts [hunk 9]
```diff
@@ -1,12 +1,12 @@
 import middy from '@middy/core'
 import { Options as MiddyOptions } from '@middy/util'
 import { Context as LambdaContext } from 'aws-lambda'
-import S3 from 'aws-sdk/clients/s3'
+import { S3Client } from '@aws-sdk/client-s3'
 import { ClientRequest } from 'http'
 
-interface Options<S = S3>
+interface Options<S = S3Client>
```
Comment (naorpeled) [inline packages/s3-object-response/index.d.ts:7]:
```suggestion
interface Options<AwsS3Client = S3Client>
```

File: packages/rds-signer/index.d.ts [hunk 10]
```diff
@@ -1,8 +1,8 @@
-import RDS from 'aws-sdk/clients/rds'
 import middy from '@middy/core'
 import { Options as MiddyOptions } from '@middy/util'
+import { Signer } from '@aws-sdk/rds-signer'
 
-interface Options<Signer = RDS.Signer> extends MiddyOptions<Signer, RDS.Types.ClientConfiguration> {}
+interface Options<S = Signer> extends MiddyOptions<S, Signer.Types.ClientConfiguration> {}
```
Comment (naorpeled) [inline packages/rds-signer/index.d.ts:5]:
```suggestion
interface Options<AwsRdsSigner = Signer> extends MiddyOptions<AwsRdsSigner, Signer.Types.ClientConfiguration> {}
```

File: packages/ws-response/index.d.ts [hunk 11]
```diff
@@ -1,11 +1,16 @@
-import ApiGatewayManagementApi from 'aws-sdk/clients/apigatewaymanagementapi'
+import { ApiGatewayManagementApiClient } from '@aws-sdk/client-apigatewaymanagementapi'
 import middy from '@middy/core'
 import { Options as MiddyOptions } from '@middy/util'
```
[the remark this answers is not in the document]
Comment (willfarrell) [reply packages/ws-response/index.d.ts:3]: Should all typedef import use type to be consistent? Is it not implied by being in the .d.ts.

[the remark this answers is not in the document]
Comment (naorpeled) [reply packages/ws-response/index.d.ts:3]: > Should all typedef import use type to be consistent? Is it not implied by being in the .d.ts.

You are correct, I didn't notice it was a .d.ts file, my bad.

File: packages/util/index.js [hunk 12]
```diff
@@ -211,8 +200,9 @@ export const normalizeHttpResponse = (request) => {
     typeof response?.body === 'undefined' &&
     typeof response?.headers === 'undefined'
   ) {
-    response = { body: response }
+    response = { statusCode: 200, body: response }
   }
+  response.statusCode ??= 500
```
[the remark this answers is not in the document]
Comment (willfarrell) [reply packages/util/index.js:205]: It does, 2 lines above. This 500 is for when the response is `undefined` or `{}` which is an error that should be handled properly.

File: packages/util/__tests__/index.js (hunk 2 above)
Comment (willfarrell) [reply packages/util/__tests__/index.js:518]: added

File: packages/util/index.js [hunk 13]
```diff
@@ -103,10 +95,7 @@ export const getInternal = async (variables, request) => {
     .filter((res) => res.status === 'rejected')
     .map((res) => res.reason)
   if (errors.length) {
-    // throw new Error('Failed to resolve internal values', { cause: errors })
-    const error = new Error('Failed to resolve internal values')
-    error.cause = errors
-    throw error
+    throw new Error('Failed to resolve internal values', { cause: errors })
```
[the remark this answers is not in the document]
Comment (willfarrell) [reply packages/util/index.js:98]: This is written to be generic on purpose. Someone could write a middleware that uses fetch not an an AWS service. It just happens only AWS Client API requests are used it right now.

**Comment by willfarrell** [comment]
@naorpeled Thanks for the code review!

File: packages/util/index.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (naorpeled) [reply packages/util/index.js:205]: after re-reading this I understand that I misunderstood this,
what you're saying is making sense.

Thanks and sorry :)

**Review by naorpeled — APPROVED** [review]

**Comment by naorpeled** [comment]
@willfarrell so what's left is only fixing the type check errors right?
If so, shall I fix them?

**Comment by willfarrell** [comment]
@naorpeled for an initial alpha release, yes.

There is also the likelyhood of more granular types between middleware discussed over at https://github.com/middyjs/middy/issues/914

Also, https://github.com/middyjs/middy/issues/905 could be applied to v3 if you're interested.

**Comment by naorpeled** [comment]
> @naorpeled for an initial alpha release, yes.
> 
> There is also the likelyhood of more granular types between middleware discussed over at #914
> 
> Also, #905 could be applied to v3 if you're interested.

I'll take #905, feel free to assign me :)

**Comment by naorpeled** [comment]
@willfarrell which other task can I take? :)

**Comment by willfarrell** [comment]
Just fixing the type check errors would be great.

**Comment by willfarrell** [comment]
@lmammino initial real world testing has been completed, ready for a review.

**Pull request merged**